File: hackerrank/array/easy/java_solutions/QueensAttack.java
package hackerrank.array.easy.java_solutions;

import java.util.*;

public class QueensAttack {

    static class SolutionMatrix {
        private static final int[][] DIRECTIONS = new int[][]{{1, 1}, {-1, -1}, {1, -1}, {-1, 1}, {1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        public static int queensAttack(int n, int k, int r_q, int c_q, List<List<Integer>> obstacles) {
            // Write your code here
            // check if boarders are not violated
            // queen is going (r + 1, c + 1), (r - 1, c - 1), (r + 1, c - 1), (r - 1, c + 1)
            Map<Integer, Set<Integer>> obstacleMatrix = convertToMatrix(obstacles, n);
            int result = 0;
            for (int[] direction : DIRECTIONS) {
                result += calculateAvailableCells(obstacleMatrix, direction, r_q, c_q, n);
            }
            return result;
        }

        private static int calculateAvailableCells(Map<Integer, Set<Integer>> obstacleMatrix, int[] direction, int row, int col, int n) {
            int result = -1;
            while (!areBoardersViolated(n, row, col) && !isObstacleFound(obstacleMatrix, row, col)) {
                result++;
                row += direction[0];
                col += direction[1];
            }
            return result;
        }

        private static Map<Integer, Set<Integer>> convertToMatrix(List<List<Integer>> obstacles, int n) {
            Map<Integer, Set<Integer>> storage = new HashMap<>();
            for (List<Integer> obstacle : obstacles) {
                storage.putIfAbsent(obstacle.get(0), new HashSet<>());
                storage.get(obstacle.get(0)).add(obstacle.get(1));
            }
            return storage;
        }

        private static boolean isObstacleFound(Map<Integer, Set<Integer>> obstacles, int row, int col) {
            if (obstacles.containsKey(row)) {
                return obstacles.get(row).contains(col);
            }
            return false;
        }

        private static boolean areBoardersViolated(int n, int row, int col) {
            return row < 1 || col < 1 || row > n || col > n;
        }
    }
}

File: hackerrank/array/easy/golang_solutions/queens_attack.go
package golang_solutions

// https://www.hackerrank.com/challenges/queens-attack-2/problem

var directions = [][]int32{{-1, -1}, {1, 1}, {-1, 1}, {1, -1}, {1, 0}, {-1, 0}, {0, 1}, {0, -1}}

func queensAttack(n int32, k int32, r_q int32, c_q int32, obstacles [][]int32) (result int32) {
	// Write your code here
	obstaclesMap := make(map[int32]map[int32]interface{})
	for _, v := range obstacles {
		if _, ok := obstaclesMap[v[0]]; !ok {
			obstaclesMap[v[0]] = make(map[int32]interface{})
		}
		obstaclesMap[v[0]][v[1]] = nil
	}
	for _, v := range directions {
		result += collectPossibleMoves(n, r_q, c_q, v, obstaclesMap)
	}
	return
}

func collectPossibleMoves(n, row, col int32, direction []int32, obstacles map[int32]map[int32]interface{}) int32 {
	result := int32(-1)
	for !isBoarderViolated(row, col, n) && !isObstacleFound(row, col, obstacles) {
		row += direction[0]
		col += direction[1]
		result++
	}
	return result
}

func isObstacleFound(row, col int32, obstacles map[int32]map[int32]interface{}) (isFound bool) {
	_, isFound = obstacles[row][col]
	return
}

func isBoarderViolated(row, col, n int32) bool {
	return row < 1 || col < 1 || row > n || col > n
}
